src/hooks/useSmartContract.tsx

```diff
@@ -97,48 +45,86 @@ export const useSmartContract = () => {
 
       console.log("Initializing provider and signer...");
-      const provider = await getProvider();
-      const signer = await provider.getSigner();
 
       console.log("Creating contract instance...");
-      const contract = await getContractInstance(contractAddress, signer);
-
-      const mintArgs: [string, bigint, boolean, string] = [
-        contextAccounts[0] as `0x${string}`,
-        ethers.parseUnits(amount.toString(), "wei"), // Amount could be ethers or wei
-        false as boolean,
-        "0x" as string,
-      ];
+      const contract = await getContractInstance(contractAddress, client);
+      console.log("contract", contract);
+      const data: string = contract.interface.encodeFunctionData("allCanMint"); // this is a custom function
 
-      const data: string = contract.interface.encodeFunctionData(
-        "mint",
-        mintArgs
-      );
-
-      // Sending the transaction via the grid provider
       console.log("Sending transaction...");
       const txResponse = await client.sendTransaction({
-        account: contextAccounts[0] as `0x${string}`,
+        account: accounts[0] as `0x${string}`,
         to: contractAddress as `0x${string}`,
         data: data,
       });
 
       console.log("Transaction response:", txResponse);
     } catch (error) {
       console.error("Transaction failed:", error);
-      throw error;
+      return;
     }
+  }, [client, walletConnected, accounts]);
+
+  const getProvider = async (): Promise<BrowserProvider> => {
+    if (!window.lukso) {
+      throw new Error("Lukso provider is not available on window object.");
+    }
+    return new ethers.BrowserProvider(window.lukso);
+  };
+
+  {
+    /**
+   This is the method to interact with the smart contract using ethers.js. While this approach is valid,
+   we recommend using the `up-provider` for interacting with the blockchain whenever possible.
+  
+   Why use `up-provider`?
+   - Simplified integration with the Universal Profile ecosystem.
+   - Enhanced security by leveraging the UP infrastructure.
+
+   With `up-provider`, the interaction is abstracted, reducing complexity and potential errors.
+   */
+  }
+
+  const getSigner = useCallback(async (): Promise<JsonRpcSigner> => {
+    const browserProvider = await getProvider();
+    return browserProvider.getSigner();
   }, []);
 
+  const getContractInstance = useCallback(
+    async (
+      contractAddress: string,
+      signer: ethers.Signer
+    ): Promise<Contract> => {
+      return new ethers.Contract(contractAddress, lsp7Json.abi, signer);
+    },
+    []
+  );
+
+  const executeFunction = useCallback(
+    async (contractAddress: string, functionName: string, params: any[]) => {
+      try {
+        const signer = await getSigner();
+        const contract = await getContractInstance(contractAddress, signer);
+        const tx = await contract[functionName](...params);
+        await tx.wait();
+        console.log("Transaction successful:", tx);
```

Comment:
We should ideally avoid `console.log` and use some debugging library like https://github.com/debug-js/debug#readme